server/src/main/java/org/opensearch/index/translog/RemoteFsTranslog.java

```diff
@@ -535,50 +613,248 @@ protected Releasable drainSync() {
+                        );
+
+                        // Delete stale metadata files
+                        translogTransferManager.deleteMetadataFilesAsync(
+                            metadataFilesToBeDeleted,
+                            remoteGenerationDeletionPermits::release
+                        );
+
+                        // Delete stale primary terms
+                        deleteStaleRemotePrimaryTerms(metadataFiles);
+                    } else {
+                        remoteGenerationDeletionPermits.release(REMOTE_DELETION_PERMITS);
+                    }
+                } catch (Exception e) {
+                    remoteGenerationDeletionPermits.release(REMOTE_DELETION_PERMITS);
+                }
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                remoteGenerationDeletionPermits.release(REMOTE_DELETION_PERMITS);
+                logger.error("Exception while listing translog metadata files", e);
+            }
+        };
+        translogTransferManager.listTranslogMetadataFilesAsync(listMetadataFilesListener);
+    }
+
+    // Visible for testing
+    protected Set<Long> getGenerationsToBeDeleted(
+        Map<String, Tuple<Long, Long>> metadataFileGenerationMap,
+        List<String> metadataFilesToBeDeleted,
+        boolean indexDeleted
+    ) throws IOException {
+
+        Set<Long> generationsToBeDeleted = new HashSet<>();
+
+        long maxGenerationToBeDeleted = minRemoteGenReferenced - 1 - indexSettings().getRemoteTranslogExtraKeep();
+
+        if (RemoteStoreSettings.isPinnedTimestampsEnabled()) {
+            String latestMetadataFileToBeDeleted = metadataFilesToBeDeleted.get(0);
+            long maxGenerationFromLatestMetadataFile = metadataFileGenerationMap.containsKey(latestMetadataFileToBeDeleted)
+                ? metadataFileGenerationMap.get(latestMetadataFileToBeDeleted).v2()
+                : getMinMaxTranslogGenerationFromMetadataFile(latestMetadataFileToBeDeleted, translogTransferManager).v2();
+
+            if (indexDeleted) {
+                maxGenerationToBeDeleted = maxGenerationFromLatestMetadataFile;
+            } else {
+                maxGenerationToBeDeleted = Math.min(maxGenerationToBeDeleted, maxGenerationFromLatestMetadataFile);
             }
-            generationsToDelete.add(generation);
         }
-        if (generationsToDelete.isEmpty() == false) {
-            deleteRemoteGeneration(generationsToDelete);
-            translogTransferManager.deleteStaleTranslogMetadataFilesAsync(remoteGenerationDeletionPermits::release);
-            deleteStaleRemotePrimaryTerms();
-        } else {
-            remoteGenerationDeletionPermits.release(REMOTE_DELETION_PERMITS);
+
+        // From the remaining files, read the oldest file to get min generation to be deleted
+        String oldestMetadataFileToBeDeleted = metadataFilesToBeDeleted.get(metadataFilesToBeDeleted.size() - 1);
```

Comment:
this might be too old in time if there are too many pinned timestamps . This would result in lots of repetitive translog deletion of same file.  Can we keep a `remoteFileTracker` as well , which can avoid this .

Reply:
But we will not delete the generation if it is pinned, right? So, it won't be deleted. With the new logic, generations and metadata files are linked. We delete generations only from the metadata files that are eligible to be deleted.

Reply:
`minGenerationToBeDeleted` can be very low and `minGenerationToBeDeleted` be very high , once a snapshot is deleted . In that case , there will be lots of `generation` , which will be accumulated in `generationsToBeDeleted` , resulting in call rate spike .

Reply:
Agree, nice catch! We will un-necessarily delete the already deleted generations. Let me make the changes.

Reply:
Changed the implmenetation.